main/ejml-core/src/org/ejml/ops/DMonoids.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright (c) 2009-2020, Peter Abeles. All Rights Reserved.
+ *
+ * This file is part of Efficient Java Matrix Library (EJML).
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.ejml.ops;
+
+/**
+ * as defined in the graphblas c-api (https://people.eecs.berkeley.edu/~aydin/GraphBLAS_API_C_v13.pdf)
+ * p. 26
+ */
+public final class DMonoids {
+    public static final DMonoid AND = new DMonoid(1, (a, b) -> (a == 0 || b == 0) ? 0 : 1);
+    public static final DMonoid OR = new DMonoid(0, (a, b) -> (a != 0 || b != 0) ? 1 : 0);
+    public static final DMonoid XOR = new DMonoid(0, (a, b) -> ((a == 0 && b == 0) || (a != 0 && b != 0)) ? 0 : 1);
+    public static final DMonoid XNOR = new DMonoid(0, (a, b) -> ((a == 0 && b == 0) || (a != 0 && b != 0)) ? 1 : 0);
+
+    public static final DMonoid PLUS = new DMonoid(0, Double::sum);
+    public static final DMonoid TIMES = new DMonoid(1, (a, b) -> a * b);
+
+    // TODO: performance incr. worth not using safe Math.min/max?
```

Comment:
This depends on the JVM and should probably be benchmarked. Also, the semantics are different. 

cf. OpenJDK implementations below:

https://github.com/openjdk/jdk/blob/93904460818dfbd6bf028b06777238603ce78614/src/java.base/share/classes/java/lang/Math.java#L1569-L1580
https://github.com/openjdk/jdk/blob/93904460818dfbd6bf028b06777238603ce78614/src/java.base/share/classes/java/lang/Math.java#L1652-L1663

cc: @lessthanoptimal

Reply:
I meant to remove the `TODO`, as I don't think I will address it. 
It was more a general comment, as I also saw that @lessthanoptimal used hard-coded min/max versions at other places.
As the library tries to avoid using unneccessary wrappers, I think using hard-coded versions is the "safe" way to go. 

I would be interested in results of those benchmarks nonetheless (probably already exist somewhere)

Thank you for pointing out, that the semantics actually differ!

Reply:
Based on my tests (OpenJDK 11.0.7 amd64), `Math.max(a, b)` is quite a bit faster than the hard-coded version. @lessthanoptimal Might want to look into that.